File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/TestSystem/WebDriverWrapper.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing.Imaging;
using System.IO;
using System.Threading;
using System.Web;
using Microsoft.Win32;
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Remote;

namespace Autotests.Utilities.WebTestCore.TestSystem
{
    public class WebDriverWrapper
    {
        private readonly IWebDriver driver;

        public WebDriverWrapper(string proxy = null)
        {
            DesiredCapabilities capabilities = DesiredCapabilities.Chrome();
            string assembliesDirectory = FindAssembliesDirectory();
            string chromePath = Path.Combine(assembliesDirectory, "Chrome", "chrome.exe");
            SetChromeVersionToRegistry(chromePath);
            var chromeOptions = new ChromeOptions
            {
                BinaryLocation = chromePath
            };
            if (!string.IsNullOrEmpty(proxy))
                chromeOptions.Proxy = new Proxy {HttpProxy = proxy};
            chromeOptions.AddArguments("start-maximized");
            chromeOptions.AddUserProfilePreference("webdriver_enable_native_events", false);
            chromeOptions.AddUserProfilePreference("download.prompt_for_download", false);
            chromeOptions.AddUserProfilePreference("download.default_directory",
                WebDriverDownloadDirectory.DirectoryPath);
            chromeOptions.AddUserProfilePreference("download.directory_upgrade", true);
            chromeOptions.AddUserProfilePreference("profile.content_settings.pattern_pairs",
                new Dictionary<string, object>
                {
                    {"http://*,*", new {plugins = 1}}
                });
            capabilities.SetCapability(ChromeOptions.Capability, chromeOptions);
            string directory = Path.Combine(assembliesDirectory, "Selenium");
            driver = new ChromeDriver(directory, chromeOptions);
        }

        public string Url
        {
            get { return driver.Url; }
        }

        public void GoToUri(Uri uri)
        {
            driver.Navigate().GoToUrl(uri.AbsoluteUri);
        }

        public void Refresh()
        {
            driver.Navigate().Refresh();
        }

        public ISearchContext GetSearchContext()
        {
            return driver;
        }
        
        public void DeleteAllCookies()
        {
            try
            {
                driver.Manage().Cookies.DeleteAllCookies();
            }
            catch (UnhandledAlertException)
            {
                var alert = driver.SwitchTo().Alert();
                alert.Dismiss();
                driver.Manage().Cookies.DeleteAllCookies();
            }
            
        }

        public void SetCookie(string cookieName, string cookieValue)
        {
            driver.Manage().Cookies.AddCookie(new Cookie(cookieName, cookieValue, "/"));
        }

//        public string FindCookie(string cookieName)
//        {
//            var cookie = driver.Manage().Cookies.GetCookieNamed(cookieName);
//            if (cookie == null)
//                return null;
//            return cookie.Value;
//        }

        public object ExecuteScript(string script, params object[] args)
        {
            return ((IJavaScriptExecutor) driver).ExecuteScript(script, args);
        }

        public void WaitForAjax()
        {
            while (true) // Handle timeout somewhere
            {
                var ajaxIsComplete = (bool)(driver as IJavaScriptExecutor).ExecuteScript("return jQuery.active == 0");
                if (ajaxIsComplete)
                    break;
            }
            while (true) // Handle timeout somewhere
            {
                var ajaxIsComplete = (bool)(driver as IJavaScriptExecutor).ExecuteScript("return typeof(Ajax) != 'function' || Ajax.activeRequestCount == 0;");
                if (ajaxIsComplete)
                    break;
            }
            while (true) // Handle timeout somewhere
            {
                            var ajaxIsComplete = (bool)(driver as IJavaScriptExecutor).ExecuteScript("return typeof(dojo) != 'function' ||  dojo.io.XMLHTTPTransport.inFlight.length == 0;");
                if (ajaxIsComplete)
                    break;
            }
        }

        public IAlert Alert(int timeout = 6000, int waitTimeout = 100)
        {
            var w = Stopwatch.StartNew();
            do
            {
                try
                {
                    return driver.SwitchTo().Alert();
                }
                catch (Exception)
                {
                    Thread.Sleep(waitTimeout);
                }
            } while (w.ElapsedMilliseconds < timeout);
            Assert.Fail(string.Format("Не смогли дождаться страницу за {0} мс", timeout));
            return driver.SwitchTo().Alert();
            
        }

        public IWebDriver SwitchToFrame(IWebElement value)
        {
            return driver.SwitchTo().Frame(value);
        }

        public IWebDriver SwitchToDefaultContent()
        {
            return driver.SwitchTo().DefaultContent();
        }

        public void Quit()
        {
            try
            {
                try
                {
                    driver.Close();
                }
                finally
                {
                    try
                    {
                        driver.Quit();
                    }
                    finally
                    {
                        driver.Dispose();
                    }
                }
            }
            catch (Exception exception)
            {
                Console.WriteLine("Ошибка при остановке ChromeDriver:\r\n{0}", exception);
            }
        }

        public void CaptureScreenshot()
        {
            Random random = new Random();
            int num = random.Next();
            String dir = Environment.CurrentDirectory;
            Directory.CreateDirectory(dir + @"\screenshots");
            String saveLocation = dir + @"\screenshots\screenshot" + num + ".png";
            try
            {
                ITakesScreenshot screenshotDriver = driver as ITakesScreenshot;
                Screenshot screenshot = screenshotDriver.GetScreenshot();
                screenshot.SaveAsFile(saveLocation, ImageFormat.Png);
                Console.WriteLine("Screenshot saved to:\r\n{0}", saveLocation);
            }
            catch (IOException)
            {

            }
        }

        public void CleanDownloadDirectory()
        {
            WebDriverDownloadDirectory.Clean();
        }

        public string[] GetDownloadedFileNames()
        {
            return WebDriverDownloadDirectory.GetDownloadedFileNames();
        }

        public void WaitDownloadFiles(int expectedCountFiles, int maximalWaitTime = 15000, int sleepInterval = 100)
        {
            WebDriverDownloadDirectory.WaitDownloadFiles(expectedCountFiles, maximalWaitTime, sleepInterval);
        }

        internal string GetUrlParameter(string parameterName)
        {
            var url = new Uri(Url);
            return HttpUtility.ParseQueryString(url.Query).Get(parameterName);
        }

        private static void SetChromeVersionToRegistry(string chromePath)
        {
            string chromeVersion = FileVersionInfo.GetVersionInfo(chromePath).ProductVersion;
            RegistryKey key =
                Registry.CurrentUser.CreateSubKey(
                    @"Software\Google\Update\Clients\{8A69D345-D564-463c-AFF1-A69D9E530F96}");
            key.SetValue("pv", chromeVersion);
            key.Close();
        }

        private string FindAssembliesDirectory()
        {
            DirectoryInfo currentDirectory = Directory.GetParent(AppDomain.CurrentDomain.BaseDirectory);
            while (true)
            {
                if (currentDirectory == null)
                    throw new Exception("The folder Assemblies not found");
                DirectoryInfo[] directories = currentDirectory.GetDirectories();
                foreach (DirectoryInfo directoryInfo in directories)
                {
                    if (directoryInfo.Name == "Assemblies" || directoryInfo.Name == "assemblies")
                        return directoryInfo.FullName;
                }
                currentDirectory = currentDirectory.Parent;
            }
        }
    }
}

File: WebDriverTests_Java/WebDriverTests/src/test/java/core/webDriver/WebDriverWrapper.java
package core.webDriver;
import com.google.common.base.Stopwatch;
import com.sun.jna.platform.unix.X11;
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.io.*;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.concurrent.TimeUnit;

public class WebDriverWrapper
{
    private WebDriver driver;
    private X11.Display display;

    public WebDriverWrapper()
    {
        DesiredCapabilities capabilities = DesiredCapabilities.chrome();
        String assembliesDirectory = FindAssembliesDirectory();
        ChromeOptions chromeOptions = new ChromeOptions();
        String OsType = System.getProperty("os.name" );
        if (OsType.startsWith("Windows"))
        {
            String directory = combine(new String[]{assembliesDirectory, "//Selenium", "//chromedriver.exe"});
            System.setProperty("webdriver.chrome.driver",directory);
            String chromePath = combine(new String[]{assembliesDirectory, "//Chrome", "//chrome.exe"});
            chromeOptions.setBinary(chromePath);

            chromeOptions.addArguments("start-maximized");
            chromeOptions.addArguments("--disable-webdriver-enable-native-events");
            chromeOptions.addArguments("--disable-download.prompt-for-download");
            chromeOptions.addArguments("--download.default-directory=", WebDriverDownloadDirectory.DirectoryPath);
            chromeOptions.addArguments("--enable-download.directory_upgrade");
            chromeOptions.addArguments("--profile-content-settings.pattern_pairs=http://*,*");
            capabilities.setCapability(ChromeOptions.CAPABILITY, chromeOptions);
        }
        else  if (OsType.startsWith("Linux")) {

                final X11 x11 = X11.INSTANCE;
                X11.Display display = x11.XOpenDisplay(null);
                X11.Window window = new X11.Window();

//            String directory = combine(new String[]{assembliesDirectory, "//Selenium", "//usr//bin//google-chrome"});
//            try {
//                Runtime.getRuntime().exec("chmod +x"+ directory);
//            } catch (IOException e) {
//                e.printStackTrace();
//            }
//            System.setProperty("webdriver.chrome.driver", directory);
            System.setProperty("webdriver.chrome.driver", "//usr//local//sbin//chromedriver");

//            String chromePath = combine(new String[]{assembliesDirectory, "//Chrome", "//usr//bin//google-chrome"});
//            try {
//                Runtime.getRuntime().exec("chmod +x"+ chromePath);
//            } catch (IOException e) {
//                e.printStackTrace();
//            }
//            chromeOptions.setBinary(chromePath);

            try {
                Runtime.getRuntime().exec("Xvfb :0 -ac -screen 0 1024x768x24 &");
            } catch (IOException e) {
                e.printStackTrace();
            }
            chromeOptions.setBinary("//usr//bin//chromium");
            chromeOptions.addArguments("--privileged");
            chromeOptions.addArguments("--no-sandbox");
            String chromePath = combine(new String[]{assembliesDirectory, "//Chrome", "//google-chrome-stable_current_x86_64.rpm"});
            chromeOptions.setBinary(chromePath);
        }
//        driver = new ChromeDriver(chromeOptions);

        String USERNAME = "andreiarnautov1";
        String AUTOMATE_KEY = "yiFysr9R61GJTRg1TDzt";
        String URL = "https://" + USERNAME + ":" + AUTOMATE_KEY + "@hub-cloud.browserstack.com/wd/hub";
        DesiredCapabilities caps = new DesiredCapabilities();
        caps.setCapability("browser", "IE");
        caps.setCapability("browser_version", "7.0");
        caps.setCapability("os", "Windows");
        caps.setCapability("os_version", "XP");
        caps.setCapability("browserstack.debug", "true");

        try {
            driver = new RemoteWebDriver(new URL(URL), caps);
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }

    }

    public static String combine(String[] args) {
        String result = "";
        for (int i=0; i<args.length; i++ )
        {
            result = result.concat(args[i]);
        }
        return result;
    }

    public final String getUrl()
    {
        return driver.getCurrentUrl();
    }

    public final void GoToUri(URL uri)
    {
        driver.navigate().to(uri);
    }

    public final void Refresh()
    {
        driver.navigate().refresh();
    }

    public final SearchContext GetSearchContext()
    {
        return driver;
    }

    public final void DeleteAllCookies()
    {
        try
        {
            driver.manage().deleteAllCookies();
        }
        catch (UnhandledAlertException e)
        {
            Alert alert = driver.switchTo().alert();
            alert.dismiss();
            driver.manage().deleteAllCookies();
        }

    }

    public final void SetCookie(String cookieName, String cookieValue)
    {
        driver.manage().addCookie(new Cookie(cookieName, cookieValue, "/"));
    }

        public String FindCookie(String cookieName)
        {
            Cookie cookie = driver.manage().getCookieNamed(cookieName);
            if (cookie == null)
                return null;
            return cookie.getValue();
        }

    public static final class DotNetToJavaStringHelper
    {
        public static String substring(String string, int start, int length)
        {
            if (length < 0)
                throw new IndexOutOfBoundsException("Parameter length cannot be negative.");

            return string.substring(start, start + length);
        }
    }

    public final Object ExecuteScript(String script)
    {
        JavascriptExecutor js = (JavascriptExecutor) driver;
        return js.executeScript(script);
    }

    public final Object javascriptExecutor(String script, WebElement webElement)
    {
        JavascriptExecutor jse = (JavascriptExecutor) driver;
        return jse.executeScript(script, webElement);
    }
    public final void WaitForAjax()
    {
        while (true) // Handle timeout somewhere
        {
            JavascriptExecutor js = (JavascriptExecutor) driver;
            boolean ajaxIsComplete = (boolean) js.executeScript("return jQuery.active == 0");
            if (ajaxIsComplete)
            {
                break;
            }
        }
        while (true) // Handle timeout somewhere
        {
            JavascriptExecutor js = (JavascriptExecutor) driver;
            boolean ajaxIsComplete = (boolean) js.executeScript("return typeof(Ajax) != 'function' || Ajax.activeRequestCount == 0;");
            if (ajaxIsComplete)
            {
                break;
            }
        }
        while (true) // Handle timeout somewhere
        {JavascriptExecutor js = (JavascriptExecutor) driver;
            boolean ajaxIsComplete = (boolean) js.executeScript("return typeof(dojo) != 'function' ||  dojo.io.XMLHTTPTransport.inFlight.length == 0;");
            if (ajaxIsComplete)
            {
                break;
            }
        }
    }

    public final Alert Alert(int timeout) {
        return Alert(timeout, 100);
    }

    public final Alert Alert() {
        return Alert(6000, 100);
    }

    public final Alert Alert(int timeout, int waitTimeout) {
        Stopwatch w = Stopwatch.createStarted();
        do
        {
            try
            {
                return driver.switchTo().alert();
            }
            catch (RuntimeException e)
            {
                try {
                    Thread.sleep(waitTimeout);
                } catch (InterruptedException e1) {
                    e1.printStackTrace();
                }
            }
        } while (w.elapsed(TimeUnit.NANOSECONDS) < timeout);
        return driver.switchTo().alert();
    }
    public final WebDriver switchToFrame(By locator)
    {
                return (new WebDriverWait(driver, 10))
            .until(ExpectedConditions.frameToBeAvailableAndSwitchToIt(locator));
//        return driver.switchTo().frame(driver.findElement(By.tagName("iframe")));
    }

    public final WebDriver switchToDefaultContent()
    {
        return driver.switchTo().defaultContent();
    }

    public final void Quit()
    {
        try
        {
            try
            {
                driver.close();
            }
            finally
            {
                driver.quit();
            }
        }
        catch (RuntimeException exception)
        {
            System.out.printf("Ошибка при остановке ChromeDriver:\r\n%1$s" + "\r\n", exception);
        }
    }



    public final void CleanDownloadDirectory()
    {
        WebDriverDownloadDirectory.Clean();
    }

    public final String[] GetDownloadedFileNames()
    {
        return WebDriverDownloadDirectory.GetDownloadedFileNames();
    }

    public final void WaitDownloadFiles(int expectedCountFiles, int maximalWaitTime) throws InterruptedException {
        WaitDownloadFiles(expectedCountFiles, maximalWaitTime, 100);
    }

    public final void WaitDownloadFiles(int expectedCountFiles) throws InterruptedException {
        WaitDownloadFiles(expectedCountFiles, 15000, 100);
    }

    public final void WaitDownloadFiles(int expectedCountFiles, int maximalWaitTime, int sleepInterval) throws InterruptedException {
        WebDriverDownloadDirectory.WaitDownloadFiles(expectedCountFiles, maximalWaitTime, sleepInterval);
    }

//    private static void SetChromeVersionToRegistry(String chromePath)
//    {
//        String chromeVersion = fileVersionInfo.GetVersionInfo(chromePath).ProductVersion;
//        RegistryKey key = Registry.currentUser.CreateSubKey("Software\\Google\\Update\\Clients\\{8A69D345-D564-463c-AFF1-A69D9E530F96}");
//        key.setValue("pv", chromeVersion);
//        key.close();
//    }

    private String FindAssembliesDirectory()
    {
//        java.io.File currentDirectory = Directory.GetParent(AppDomain.CurrentDomain.BaseDirectory);
        java.io.File currentDirectory = new File(".").getAbsoluteFile().getParentFile();
        while (true)
        {
            if (currentDirectory == null)
            {
                throw new RuntimeException("The folder Assemblies not found");
            }
            java.io.File[] directories = currentDirectory.listFiles();
            for (java.io.File directoryInfo : directories)
            {
                if (directoryInfo.getName().equals("Assemblies") || directoryInfo.getName().equals("assemblies"))
                {
                    return directoryInfo.getPath();
                }
            }
            currentDirectory = currentDirectory.getParentFile();
        }
    }

    public byte[] CaptureScreenshot()
    {
        return ((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES);
//        Random random = new Random();
//        int num = random.nextInt();
//        try {
//            WebDriver returned = new Augmenter().augment(driver);
//            if (returned != null) {
//                File f = ((TakesScreenshot) returned)
//                        .getScreenshotAs(OutputType.FILE);
//                try {
//                    FileUtils.copyFile(f, new File("target\\allure-results" + name + ".jpg"));
//                } catch (IOException e) {
//                    e.printStackTrace();
//                }
//            }
//        } catch (ScreenshotException se) {
//            se.printStackTrace();
//        }
    }
}


